### PR TITLE
sqldb: fix timeparsing edge cases

### PR DESCRIPTION
**Pull request by sputn1ck**
This PR adds additional fixes for more timestamp formats.

I was not able to reproduce https://github.com/lightninglabs/loop/issues/628, but I added improved logging, so should the error persist we would be more able to fully grasp the failed string.

### CODE REVIEW

File: loopdb/sqlite.go [hunk 1]
```diff
@@ -377,12 +378,25 @@ func parseTimeParts(datePart, timePart string) (time.Time, error) {
 		return time.Time{}, err
 	}
 
-	day, err := strconv.Atoi(dateParts[2])
+	dateIndex := 2
```
Comment (hieblmi) [inline loopdb/sqlite.go:381]: I find it hard to parse what the code here is doing. Maybe we can add the template date string that is handled here to clarify?

Comment (hieblmi) [reply loopdb/sqlite.go:381]: Also, maybe in the two returns above maybe we can add `unable to atoi month` and `unable to atoi year` to the error string.

**Review by hieblmi — COMMENTED** [review]
I am thinking that rather than adjusting the date parsing manually for every new issue we could maybe use a battle tested date parsing module here?

**Review by hieblmi — APPROVED** [review]
lgtm ✅

File: loopdb/sqlite.go [hunk 2]
```diff
@@ -369,20 +370,27 @@ func parseTimeParts(datePart, timePart string) (time.Time, error) {
...
 	day, err := strconv.Atoi(dateParts[2])
 	if err != nil {
-		return time.Time{}, err
+		return time.Time{}, fmt.Errorf("unable to parse day: %v", err)
 	}
 
 	// Parse the time.
+
+	// Remove the timezone part if it exists. E.g. 2023-08-04 16:07:49+0800 CST
+	if strings.Contains(timePart, "+") {
```
Comment (bhandras) [inline loopdb/sqlite.go:389]: I'm afraid there's some more things to consider. I was trying to look up where these things come from and it seems to me that this is governed by the ISO 8601 standard (https://en.wikipedia.org/wiki/ISO_8601). According to that specification the offset may also be negative.

Comment (GeorgeTsagk) [reply loopdb/sqlite.go:389]: +1

**Review by GeorgeTsagk — APPROVED** [review]
Lgtm, pending the timezone nit

File: loopdb/sqlite.go [hunk 3]
```diff
@@ -335,9 +336,13 @@ func parseSqliteTimeStamp(dateTimeStr string) (time.Time, error) {
...
 
-	datePart, timePart := parts[0], parts[1]
+	dateTimeStr, err := replaceFaultyYear(parts[0], dateTimeStr)
+	if err != nil {
+		return time.Time{}, fmt.Errorf("%v: unable to replace faulty "+
+			"year: %v", dateTimeStr, err)
+	}
 
-	return parseTimeParts(datePart, timePart)
+	return time.Parse("2006-01-02 15:04:05.99999 -0700 MST", dateTimeStr)
```
Comment (bhandras) [inline loopdb/sqlite.go:345]: Instead of this, I'm wondering if we should try to instead parse the time based on the supported RFC's as documented here: https://pkg.go.dev/time#pkg-constants

File: loopdb/sqlite.go [hunk 4]
```diff
@@ -311,117 +312,87 @@ func (r *SqliteTxOptions) ReadOnly() bool {
...
+		)
 	}
 
-	datePart, timePart := parts[0], strings.TrimSuffix(parts[1], "Z")
+	var layouts []string
 
-	return parseTimeParts(datePart, timePart)
-}
+	// Add the custom go sqlite layout.
+	layouts = append(layouts, "2006-01-02 15:04:05.99999 -0700 MST")
```
Comment (bhandras) [inline loopdb/sqlite.go:341]: nit: maybe we could add this layout to the `DefaultLayouts` function too.

File: loopdb/sqlite.go [hunk 5]
```diff
@@ -311,117 +312,87 @@ func (r *SqliteTxOptions) ReadOnly() bool {
...
+func DefaultAllLayouts() []string {
+	return []string{
+		time.RFC3339Nano,
+		time.RFC3339,
+		time.RFC1123Z,
+		time.RFC1123,
+		time.RFC850,
+		time.RFC822Z,
+		time.RFC822,
+		time.Layout,
```
Comment (bhandras) [inline loopdb/sqlite.go:378]: So I wonder if we need all these layouts?

File: loopdb/sql_test.go [hunk 6]
```diff
@@ -516,3 +519,68 @@ func randomStruct(v interface{}) error {
 
 	return nil
 }
+
+// parseTimeParts takes a datePart string in the format of "YYYY-MM-DD" and
+// a timePart string in the format of "HH:MM:SS" and returns a time.Time value.
+func parseTimeParts(dateTimeStr string) (time.Time, error) {
```
Comment (bhandras) [inline loopdb/sql_test.go:525]: Do we need this function?

**Review by bhandras — APPROVED** [review]
Fix LGTM, just a few nits! 🎉
